src/hotspot/share/prims/upcallLinker.cpp

```diff
@@ -72,6 +72,10 @@ JavaThread* UpcallLinker::on_entry(UpcallStub::FrameData* context) {
   // clear any pending exception in thread (native calls start with no exception pending)
   thread->clear_pending_exception();
 
+  // The call to transition_from_native below contains a safepoint check
+  // which needs the code cache to be writable.
+  MACOS_AARCH64_ONLY(ThreadWXEnable wx(WXWrite, thread));
```

Comment:
Naive question: should we switch back to non-writeable after the transition?

Reply:
That happens when the `wx` variable goes out of scope: `ThreadWXEnable` is a RAII object which saves the previous mode in its constructor and then restores in the destructor.

Reply:
Yes it doesn't matter that it's writable longer than strictly necessary, as long as we make it executable again before returning to the stub. We enable the W mode around the whole body of other VM entries - e.g. see `VM_LEAF_BASE` in `interfaceSupport.inline.hpp`.